File: ios-native/FixItLens/Services/APIClient.swift
import Foundation
import UIKit

@MainActor
final class APIClient {
    static let shared = APIClient()
    private let session: URLSession
    private let timeout: TimeInterval = 30

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        session = URLSession(configuration: config)
    }

    private func baseURL() throws -> URL {
        let raw = AppSettings.shared.apiBaseURL.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let url = URL(string: raw) else { throw APIError.invalidURL }
        return url
    }

    private func request<T: Decodable>(_ path: String, method: String = "GET", body: Data? = nil, contentType: String? = nil) async throws -> T {
        let url = try baseURL().appendingPathComponent(path.trimmingCharacters(in: CharacterSet(charactersIn: "/")))
        var req = URLRequest(url: url)
        req.httpMethod = method
        if let body {
            req.httpBody = body
            req.setValue(contentType ?? "application/json", forHTTPHeaderField: "Content-Type")
        }
        do {
            let (data, response) = try await session.data(for: req)
            guard let http = response as? HTTPURLResponse else {
                throw APIError.network("No response from server.")
            }
            if http.statusCode >= 400 {
                if let err = try? JSONDecoder().decode(APIErrorResponse.self, from: data) {
                    throw APIError.server(err.detail)
                }
                throw APIError.server("Request failed (\(http.statusCode)).")
            }
            do {
                return try JSONDecoder().decode(T.self, from: data)
            } catch {
                throw APIError.decoding
            }
        } catch let e as APIError {
            throw e
        } catch {
            throw APIError.network("Could not reach FixIt Lens backend. Check Account → Backend connection and ensure the backend is running.")
        }
    }

    func health() async throws -> HealthResponse {
        try await request("/health")
    }

    func analyzeImage(_ imageData: Data, filename: String = "capture.jpg", userHint: String? = nil) async throws -> AnalyzeImageResponse {
        let boundary = UUID().uuidString
        var body = Data()
        func append(_ s: String) { body.append(Data(s.utf8)) }

        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"image\"; filename=\"\(filename)\"\r\n")
        append("Content-Type: image/jpeg\r\n\r\n")
        body.append(imageData)
        append("\r\n")
        if let hint = userHint, !hint.isEmpty {
            append("--\(boundary)\r\n")
            append("Content-Disposition: form-data; name=\"user_hint\"\r\n\r\n")
            append("\(hint)\r\n")
        }
        append("--\(boundary)--\r\n")

        let url = try baseURL().appendingPathComponent("api/analyze/image")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        req.httpBody = body

        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse, http.statusCode < 400 else {
            if let err = try? JSONDecoder().decode(APIErrorResponse.self, from: data) {
                throw APIError.server(err.detail)
            }
            throw APIError.server("Image analysis failed.")
        }
        return try JSONDecoder().decode(AnalyzeImageResponse.self, from: data)
    }

    func diagnose(_ payload: DiagnoseRequest) async throws -> DiagnoseResponse {
        let body = try JSONEncoder().encode(payload)
        return try await request("/api/diagnose", method: "POST", body: body)
    }

    func listSessions() async throws -> [SessionSummary] {
        try await request("/api/sessions")
    }

    func deleteAllSessions() async throws {
        let _: StatusResponse = try await request("/api/sessions", method: "DELETE")
    }

    func getSession(_ id: String) async throws -> SessionDetail {
        try await request("/api/sessions/\(id)")
    }

    func submitFeedback(sessionId: String, stepNumber: Int?, result: FeedbackResult) async throws {
        let payload = FeedbackRequest(session_id: sessionId, step_number: stepNumber, result: result.rawValue, comment: nil)
        let body = try JSONEncoder().encode(payload)
        let _: StatusResponse = try await request("/api/feedback", method: "POST", body: body)
    }

    func getMetrics() async throws -> MetricsResponse {
        try await request("/api/metrics")
    }

    func uploadManual(text: String, title: String?, category: String?, brand: String?, model: String?) async throws -> ManualUploadResponse {
        let boundary = UUID().uuidString
        var body = Data()
        func append(_ s: String) { body.append(Data(s.utf8)) }
        func field(_ name: String, _ value: String) {
            append("--\(boundary)\r\n")
            append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n")
            append("\(value)\r\n")
        }
        field("text", text)
        if let title { field("title", title) }
        if let category { field("category", category) }
        if let brand { field("brand", brand) }
        if let model { field("model", model) }
        append("--\(boundary)--\r\n")

        let url = try baseURL().appendingPathComponent("api/manuals/upload")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        req.httpBody = body

        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse, http.statusCode < 400 else {
            if let err = try? JSONDecoder().decode(APIErrorResponse.self, from: data) {
                throw APIError.server(err.detail)
            }
            throw APIError.server("Manual upload failed.")
        }
        return try JSONDecoder().decode(ManualUploadResponse.self, from: data)
    }
}

struct StatusResponse: Codable {
    let status: String
}

File: mobile/src/services/apiClient.ts
import { API_BASE_URL, REQUEST_TIMEOUT_MS } from './config';
import { getApiBaseUrlOverride } from './storage';
import type {
  AnalyzeImageResponse,
  DiagnoseRequestPayload,
  DiagnoseResponse,
  FeedbackPayload,
  HealthResponse,
  ManualUploadResponse,
  MetricsResponse,
  SessionDetail,
  SessionSummary,
} from './types';

export class ApiError extends Error {
  status?: number;
  constructor(message: string, status?: number) {
    super(message);
    this.name = 'ApiError';
    this.status = status;
  }
}

async function resolveBaseUrl(): Promise<string> {
  const override = await getApiBaseUrlOverride();
  return override && override.trim().length > 0 ? override.trim() : API_BASE_URL;
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const baseUrl = await resolveBaseUrl();
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);

  try {
    const response = await fetch(`${baseUrl}${path}`, { ...init, signal: controller.signal });
    if (!response.ok) {
      let detail = `Request failed with status ${response.status}`;
      try {
        const body = await response.json();
        if (body?.detail) detail = body.detail;
      } catch {
        // ignore non-JSON error bodies
      }
      throw new ApiError(detail, response.status);
    }
    return (await response.json()) as T;
  } catch (err) {
    if (err instanceof ApiError) throw err;
    if (err instanceof Error && err.name === 'AbortError') {
      throw new ApiError('The request timed out. Check that the backend is running and reachable.');
    }
    throw new ApiError('Could not reach the FixIt Lens backend. Check your connection and API URL in Settings.');
  } finally {
    clearTimeout(timeout);
  }
}

export async function getHealth(): Promise<HealthResponse> {
  return request<HealthResponse>('/health');
}

export async function analyzeImage(
  imageUri: string,
  filename: string,
  mimeType: string,
  userHint?: string
): Promise<AnalyzeImageResponse> {
  const formData = new FormData();
  formData.append('image', {
    uri: imageUri,
    name: filename,
    type: mimeType,
  } as unknown as Blob);
  if (userHint) formData.append('user_hint', userHint);

  return request<AnalyzeImageResponse>('/api/analyze/image', {
    method: 'POST',
    body: formData,
  });
}

export async function diagnose(payload: DiagnoseRequestPayload): Promise<DiagnoseResponse> {
  return request<DiagnoseResponse>('/api/diagnose', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
}

export async function uploadManualText(
  text: string,
  title?: string,
  category?: string,
  brand?: string,
  model?: string
): Promise<ManualUploadResponse> {
  const formData = new FormData();
  formData.append('text', text);
  if (title) formData.append('title', title);
  if (category) formData.append('category', category);
  if (brand) formData.append('brand', brand);
  if (model) formData.append('model', model);

  return request<ManualUploadResponse>('/api/manuals/upload', {
    method: 'POST',
    body: formData,
  });
}

export async function listSessions(): Promise<SessionSummary[]> {
  return request<SessionSummary[]>('/api/sessions');
}

export async function getSession(sessionId: string): Promise<SessionDetail> {
  return request<SessionDetail>(`/api/sessions/${sessionId}`);
}

export async function submitFeedback(payload: FeedbackPayload): Promise<{ status: string }> {
  return request<{ status: string }>('/api/feedback', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
}

export async function getMetrics(): Promise<MetricsResponse> {
  return request<MetricsResponse>('/api/metrics');
}
